### PR TITLE
Implements Accounts Listing

### PR DESCRIPTION
**Pull request by josemigallas**
**What this PR does / why we need it**:

Implements "Accounts > Listing" page for Portafly.

**Which issue(s) this PR fixes** 

[THREESCALE-4732: Create component for Account listing page](https://issues.redhat.com/browse/THREESCALE-4732)

**Verification steps** 
* To work with mock data, open a separate terminal run `yarn mock-server`.
* To work with real data from preview or your local cluster, create an access token there and use it to log in Portafly.

Once the data source is sorted out, simply run `yarn start`

**Future steps** 
This PR focuses on the Table logic, there are still a bunch of tasks and improvements related to this page (see [Epic: THREESCALE-4724](https://issues.redhat.com/browse/THREESCALE-4724)).

### CODE REVIEW

File: portafly/src/tests/App.test.tsx [hunk 1]
```diff
@@ -4,5 +4,5 @@ import { render } from 'tests/custom-render'
 
 it('renders without crashing', () => {
   const wrapper = render(<App />)
-  expect(wrapper).not.toBeNull()
+  expect(wrapper.container.firstChild).not.toBeNull()
```
Comment (damianpm) [inline portafly/src/tests/App.test.tsx:7]: Maybe this kind of improvements can be moved to a separate PR

File: portafly/src/components/developer-accounts/modals/BaseModal.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import React, { useState, useMemo } from 'react'
```
Comment (damianpm) [inline portafly/src/components/developer-accounts/modals/BaseModal.tsx:1]: Do you think all modals (or at least BaseModal) can be a separate PR?

Comment (josemigallas) [reply portafly/src/components/developer-accounts/modals/BaseModal.tsx:1]: I think the idea is to create and merge smaller PRs, having this one as a reference only so I wouldn't be worried about it.

**Comment by josemigallas** [comment]
> Why is this happening?

That's the mock server generating a dynamic example. I'm not sure how but it must be possible to disable that feature.

File: portafly/src/tests/App.test.tsx (hunk 1 above)
Comment (josemigallas) [reply portafly/src/tests/App.test.tsx:7]: Certainly

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import { PlusCircleIcon, ExportIcon } from '@patternfly/react-icons'
+
+const AccountsListingPage: React.FunctionComponent = () => {
+  const { data: accounts, error, isPending } = useAsync(getDeveloperAccounts)
+  const { t } = useTranslation('audienceAccountsListing')
+  useDocumentTitle(t('title_page'))
+
+  const isMultitenant = false // TODO: get this somehow
+
+  const DataListTable = ({ accounts: _accounts }: any) => {
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:26]: `DataListTable` can be moved outside `AccountsListingPage`

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  // FIXME: this throws an error
+  // const filteredRows = useMemo(() => (
+  //   filterRows(rows, filters, categories, columns)
+  // ), [rows, filters, sortBy])
+
+  const filteredRows = filterRows(rows, filters, columns)
+
+  const visibleRows = filteredRows.slice(startIdx, endIdx)
+
+  const onSort: OnSort = (_event, index, direction) => {
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:77]: Is this `_event` needed ?

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    setSortBy(index, direction, true)
+  }
+
+  const onSelectOne: OnSelect = (_ev, selected, _rowIndex, rowData) => {
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:81]: Same here, is `_ev` needed?
Maybe is a PF stuff?

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  const pagination = <PaginationWidget itemCount={filteredRows.length} />
+
+  const searchEmptyState = [{
+    heightAuto: true,
+    cells: [{
+      props: { colSpan: 8 },
+      title: <TableEmptyState
+        // title={t('accounts_table.filter_empty_state_title')}
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:92]: Delete comments, please

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        </ToolbarContent>
+      </Toolbar>
+
+      {modal === 'sendEmail' && (
+        <SendEmailModal
+          items={selectedRows.map((r) => `${(r.cells as string[])[1]} (${(r.cells as string[])[0]})`)}
+        />
+      )}
+
+      {modal === 'changeState' && (
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:154]: Maybe this `sendEmail`  and `changeState` can be moved to a CONSTANT

File: portafly/src/components/pages/accounts/listing/ActionButtonImpersonate.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
+import { useTranslation } from 'react-i18next'
+import { Button } from '@patternfly/react-core'
+import { BoltIcon } from '@patternfly/react-icons'
+
+const ActionButtonImpersonate = () => {
+  const { t } = useTranslation('audienceAccountsListing')
+
+  const onClick = () => {
+    // TODO
```
Comment (damianpm) [inline portafly/src/components/pages/accounts/listing/ActionButtonImpersonate.tsx:10]: Todo what?

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 4 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:77]: It's not used but it's part of the method's signature, so it has to be declared.

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 6 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:92]: These are strings missing from the strings files, it's there as reminder. I'll ping the docs team about it.

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 7 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:154]: `modal` is already typed as `BulkAction` and this type is inferred to `changeState` and `sendEmail`. If we compared anything different, it would throw an error.
What I mean is thanks to Typescript there's no need for Enum or constants.
Screen Shot 2020-06-09 at 11 12 55

File: portafly/src/components/pages/accounts/listing/ActionButtonImpersonate.tsx (hunk 8 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/ActionButtonImpersonate.tsx:10]: Todo the onClick for the "Impersonate" action.

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    { name: 'suspended', humanName: t('actions_filter_options.by_state_options.suspended') }
+  ]
+
+  const categories = [
+    { name: 'group', humanName: t('accounts_table.group_header') },
+    { name: 'admin', humanName: t('accounts_table.admin_header') },
+    { name: 'state', humanName: t('accounts_table.state_header'), options }
+  ]
+
+  // FIXME: this throws an error
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:60]: I think I'll remove those and add a ticket to the epic

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx (hunk 3 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:26]: Well, DataListTable is simply a wrap around the Provider and a way to render it only when `accounts !== undefined`. You could see it as a `const` but instead if benefits from React's nomenclature. It's just a way to keep [line 75](https://github.com/3scale/porta/pull/1820/files#diff-db85736b3e028a1b4f2e1a41af362af2R72) clean:
```jsx
{accounts && <DataListTable accounts={accounts} />}
```

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import { useTranslation } from 'i18n/useTranslation'
+import { PlusCircleIcon, ExportIcon } from '@patternfly/react-icons'
+import { IDeveloperAccount } from 'types'
+
+const AccountsListingPage: React.FunctionComponent = () => {
+  const { data: accounts, error, isPending } = useAsync(getDeveloperAccounts)
+  const { t } = useTranslation('audienceAccountsListing')
+  useDocumentTitle(t('title_page'))
+
+  const isMultitenant = false // TODO: get this somehow
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:25]: By ENV on init

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    const initialState = {
+      table: { columns, rows }
+    }
+
+    return (
+      <DataListProvider initialState={initialState}>
+        <AccountsListingTable />
+      </DataListProvider>
+    )
+  }
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:40]: This could be extracted outside `AccountsListingPage`

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
+import React, { useEffect } from 'react'
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:1]: Not entirely convinced about the path and nomenclature `listing`, maybe this could be `list` or `collection`. This applies to all of the paths and name of files

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+  const {
+    columns,
+    rows,
+    sortBy,
+    setSortBy,
+    selectedRows,
+    selectOne
+  } = useDataListTable()
+
+  if (rows.length === 0) {
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:38]: Not mandatory and is still readable,  but the `length` method will always return integer, and `0` is `falsey`, you could force the boolean value by `!!`

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    { name: 'pending', humanName: t('actions_filter_options.by_state_options.pending') },
+    { name: 'rejected', humanName: t('actions_filter_options.by_state_options.rejected') },
+    { name: 'suspended', humanName: t('actions_filter_options.by_state_options.suspended') }
+  ]
+
+  const categories = [
+    { name: 'group', humanName: t('accounts_table.group_header') },
+    { name: 'admin', humanName: t('accounts_table.admin_header') },
+    { name: 'state', humanName: t('accounts_table.state_header'), options }
+  ]
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:57]: I think it's good to have these options here, but will want to make sure these won't change give an account plan or some modifier. If that's the case, we will need to get them from the API

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+      {modal === 'sendEmail' && (
+        <SendEmailModal
+          items={selectedRows.map((r) => `${(r.cells as string[])[1]} (${(r.cells as string[])[0]})`)}
+        />
+      )}
+
+      {modal === 'changeState' && (
+        <ChangeStateModal
+          items={selectedRows.map((r) => `${(r.cells as string[])[0]} (${(r.cells as string[])[4]})`)}
```
Comment (didierofrivia) [inline portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:130]: I find it a bit cryptic the indexes of the arrays `0` and `4`

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx (hunk 10 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:25]: #1956

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx (hunk 11 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx:40]: See my response to Damian's same comment

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 12 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:1]: I strongly suggest we cope with that in another PR, perhaps of #1941

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 13 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:38]: I don't understand. The point of this predicate is to compare one integer to another.

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 15 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:130]: It is super cryptic... I've been wanting to do something like:
```jsx
`${(r.cells.group} (${r.cells.admin})`
```
I guess we could give those indexes names but.. I would leave it for the next iteration (it would involve changes in the sorting and table data factory at least).

Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:130]: We could wrap the cells into a AccountRow entity, for the sake of readability, but we would just be moving the "crypticity" somewhere else.
```jsx
const { group, admin } = new Row<IDeveloperAccount>(r.cells)
```

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 13 above)
Comment (didierofrivia) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:38]: just pointing out that `rows.length` is either `0` => falsey or any integer => `truthy`

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 15 above)
Comment (didierofrivia) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:130]: Hide it behind a fn with a name that tell us the story is good enough

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx (hunk 3 above)
Comment (didierofrivia) [reply portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx]: Having it outside will benefit of a better garbage collection. If `accounts` is falsey, then `DataListTable` will be removed completely (when outside). Either way, it stays in the closure

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 12 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:1]: #1960

File: portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx (hunk 13 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:38]: Ah! gotcha.

Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingTable.tsx:38]: So, between `!rows.length` and `rows.length === 0`, since there's already a logical operatior involved I would choose the second.

File: portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx (hunk 3 above)
Comment (josemigallas) [reply portafly/src/components/pages/accounts/listing/AccountsListingPage.tsx]: OK
